### DISCUSSION TITLE
Serving each subsite from a different server w/o using IF ELSE if possible

### DISCUSSION THREAD

**Discussion by easaw** [stuttter/ludicrousdb#123] [Q&A] [answered]
I would like to serve each sub-domain from a different server (data protection reasons).

I am looking at using LudicrousDB over the other plugins, because I am worried about PHP7 (and 8), MySQL and other compatibilities in the future (and for example I use MariaDB).

Also, if possible I would prefer to have each blog go to the right DB without using some IF statement.
Anyone know of a way to send the query to the right DB using some sort of cached info, to save time/ cycles etc.?

So far I have this:

```/** Sample Configuration 2: Partitioning **/

/**
 * This example shows a setup where the multisite blog tables have been
 * separated from the global dataset.
 */

$wpdb->add_database(array(
	'host'     => 'localhost',
	'user'     => 'global',
	'password' => 'globalPassword',
	'name'     => 'global',
));
$wpdb->add_database(array(
	'host'     => 'localhost',
	'user'     => 'home-user',
	'password' => 'homePassword',
	'name'     => 'homedb',
	'dataset'  => 'home',
));
$wpdb->add_database(array(
	'host'     => '192.168.0.2,
	'user'     => 'demo-user2',
	'password' => 'demoPassword2',
	'name'     => 'demo2',
	'dataset'  => 'demo2',
));
$wpdb->add_database(array(
	'host'     => '192.168.0.3',
	'user'     => 'demo-user3',
	'password' => 'demoPassword3',
	'name'     => 'demo3',
	'dataset'  => 'demo3',
));

$wpdb->add_callback('resolve_with_dataset');
function resolve_with_dataset ($query, $wpdb) {
  // Multisite blog tables are "{$base_prefix}{$blog_id}_*"
  if ( preg_match("/^{$wpdb->base_prefix}\d+_/i", $wpdb->table) ) {
    // break down into just wp_86_ prefix
    $prefix_matches = array();
    preg_match("/^{$wpdb->base_prefix}\d+_/i", $wpdb->table, $prefix_matches);
    $site_table_prefix = $prefix_matches[0];
    $site_id_matches = array();
    // pull out any numerical matches here
    preg_match("/\d+/i", $site_table_prefix, $site_id_matches);
    $site_id = $site_id_matches[0];
    if ($site_id == 1) {
      return 'home';
    }
    else if ($site_id == 2)
    {
      return 'demo2';
    }
    else if ($site_id == 3)
    {
      return 'demo3';
    }
    else {
    return null; //is this right? what about the global?
  }
}
```

Would this be the correct kind of thing to do (it's untested), or should I look at using SharDB?

If Yes or No, please explain so I can understand further. Thank you.

**Comment by danfeeley** [comment]
Hey easaw, did you get this to work?

**Comment by JJJ** [comment, marked as answer]
Your callback is the right general mechanism, but I would separate the site-to-dataset mapping from the SQL routing.

Parse the blog ID from `$wpdb->table`, look up a logical dataset name, and return that name. The lookup can be a deployment-generated PHP array or a non-database-backed persistent cache such as Redis or Memcached. Do not use a cache implementation that queries `$wpdb` from inside this callback, because that can recurse back into database routing.

```php
$wpdb->add_callback(
	function ( $query, $wpdb ) {
		$pattern = '/^' . preg_quote( $wpdb->base_prefix, '/' ) . '(\d+)_/';

		if ( ! preg_match( $pattern, $wpdb->table, $matches ) ) {
			return null;
		}

		$blog_id = (int) $matches[1];
		$map     = array(
			2 => 'demo2',
			3 => 'demo3',
		);

		return $map[ $blog_id ] ?? null;
	}
);
```

For a large or changing network, generate or cache that map instead of growing an `if / elseif` chain. Each returned dataset still needs a matching `add_database()` definition, and the complete site table set must already exist there. LudicrousDB routes queries; it does not provision servers or move data.

Since data isolation is the reason for the design, use separate least-privilege database credentials as well. Application routing alone is not a security boundary if one credential can read every site's database.
